ksqldb-engine/src/main/java/io/confluent/ksql/planner/LogicalPlanner.java

```diff
@@ -575,6 +579,113 @@ private JoinNode buildJoin(final Join root, final String prefix, final boolean i
+        // at any level in the join tree, even after we add right-deep/bushy join tree support,
+        // because a FK-join output table has the same PK as its left input table
+        throw new KsqlException("Invalid join condition:"
+            + " foreign-key table-table joins are not supported.");
+      }
+    }
+  }
+
+  private boolean joinOnNonKeyAttribute(final Expression joinExpression,
+                                        final PlanNode node) {
+    if (!(joinExpression instanceof ColumnReferenceExp)) {
+      return true;
+    }
+    final ColumnReferenceExp simpleJoinExpression = (ColumnReferenceExp) joinExpression;
+
+    final String joinAttributeName = simpleJoinExpression.getColumnName().text();
+    final List<DataSourceNode> dataSourceNodes = node.getSourceNodes().collect(Collectors.toList());
+    final String singleAttributeKeyName;
+
+    // n-way join sub-tree (ie, not a leaf)
+    if (isInnerNode(node)) {
+      final DataSourceNode qualifiedNode;
+
+      if (simpleJoinExpression.maybeQualifier().isPresent()) {
+        final String qualifier = simpleJoinExpression.maybeQualifier().get().text();
+        final List<DataSourceNode> allNodes = dataSourceNodes.stream()
+            .filter(n -> n.getDataSource().getName().text().equals(qualifier))
+            .collect(Collectors.toList());
+
+        if (allNodes.size() != 1) {
+          // double check this
+          throw new KsqlException(String.format(
+              "Join qualifier '%s' could not be resolved (either not found or not unique).",
+              qualifier
+          ));
+        }
+        qualifiedNode = allNodes.get(0);
+      } else {
+        final List<DataSourceNode> allNodes = dataSourceNodes.stream()
+            .filter(n -> n.getSchema().columns().stream()
+                .map(column -> column.name().text()).collect(Collectors.toList())
+                .contains(joinAttributeName))
+            .collect(Collectors.toList());
+
+        if (allNodes.size() != 1) {
+          // double check this
+          throw new KsqlException(String.format(
+              "Join identifier '%s' could not be resolved (either not found or not unique).",
+              joinAttributeName
+          ));
+        }
+        qualifiedNode = allNodes.get(0);
+      }
+
+      // we only support joins on single attributes so the key has a single field
+      singleAttributeKeyName = qualifiedNode.getSchema().key().get(0).name().text();
+    } else {
+      // leaf node:
+      // - we know we have single data source
+      // - we only support joins on single attributes so the key has a single field
+      singleAttributeKeyName = dataSourceNodes.get(0).getSchema().key().get(0).name().text();
+    }
+
+
+    return !joinAttributeName.equals(singleAttributeKeyName);
+  }
+
+  private boolean isInnerNode(final PlanNode node) {
+    return node.getSourceNodes().count() > 1;
```

Comment:
I'm still uncomfortable with the fact that this logic is brittle. If one day we decide to allow self-joins, then this breaks and it's very plausible that the person implementing self-joins won't realize they'll have broken the logic for detecting FK joins. Let's continue the discussion in https://github.com/confluentinc/ksql/pull/7452#discussion_r625262567.